### PR TITLE
Tests: run search tests by default

### PR DESCRIPTION
**Pull request by stsewd**
Now that docker is the recommended installation method
we don't need to worry that users would need to have ES installed
to run the tests.

### CODE REVIEW

File: docs/development/tests.rst
Comment (humitos) [inline docs/development/tests.rst:36]: Our `pytest` command already have `-m "not proxito"`. I think that passing another `-m` argument makes this to break. Did you test it? I'm getting this error:

```
root@a1039fd19be4:/usr/src/app/checkouts/readthedocs.org# tox -e py36 -- -m 'not search' -x

...

ERROR: Wrong expression passed to '-m': not: at column 4: expected not OR left parenthesis OR identifier; got end of input

ERROR: InvocationError for command /bin/sh -c ' export DJANGO_SETTINGS_MODULE=readthedocs.settings.test; pytest --cov-report= --cov-config /usr/src/app/checkouts/readthedocs.org/.coveragerc --cov=. --suppress-no-test-exit-code -m "not proxito" -m not' 'search -x' (exited with code 4)
```

File: docs/development/tests.rst
Comment (humitos) [inline docs/development/tests.rst:30]: Are these directions correct? It seems that we are telling people to run `tox` from _outside_ the container, but we don't really support that. I think we should update the whole guide.

**Review by humitos — COMMENTED** [review]
I'm fine running search tests. However, we should find a way to make it work when trying to skip them.

Also, we should rewrite the whole page about tests guiding the user to run the tests from _inside_ the container, since as it is written right now it doesn't work.

On the other hand, since we are running tests against just one Python version (3.6) we should probably adapt our tox suite case to be: `tox -e web`, `tox -e proxito`, `tox -e search`, and `tox -e all` so it's easier to know what tests we are running and easier to skip the ones that you don't want to run, instead of dealing with multiple broken `-m` arguments.

**Comment by stsewd** [comment]
> be: tox -e web, tox -e proxito, tox -e search, and tox -e all so it's easier to know what tests we are running and easier to skip the ones that you don't want to run, instead of dealing with multiple broken -m arguments.

The problem with that is that it will create a different environment and install all the deps for each one, that takes a lot of time.

File: docs/development/tests.rst
Comment (stsewd) [reply docs/development/tests.rst:36]: the next -m will override the former, don't remember if this was working or if I ever used it. Looks like it's ignoring the quotes and reading the command as `-m not  search`

Comment (stsewd) [reply docs/development/tests.rst:36]: I'm not really sure if we should include the search mark to be honest, I think if you want to run just a set of tests you can just use `-k`

**Comment by ericholscher** [comment]
I don't have great feedback here, but I'm mostly worried about the team running tests locally. I almost always just run a subset of tests locally (`tox -e py3 -- -s -k test_front_slash_url  readthedocs/proxito/tests/test_middleware.py`) -- so the defaults don't matter much to me.

I'd say lets aim to make things _much simpler_ by default (run all tests). The goal of speeding them in locally I imagine it going to be including instead of excluding, and then the full suite will generally only run in CI. At least that's my view of how it should be structures.

**Comment by ericholscher** [comment]
> > be: tox -e web, tox -e proxito, tox -e search, and tox -e all so it's easier to know what tests we are running and easier to skip the ones that you don't want to run, instead of dealing with multiple broken -m arguments.
> 
> The problem with that is that it will create a different environment and install all the deps for each one, that takes a lot of time.

There must be a way to reuse envs. Does this work? https://blog.ionelmc.ro/2015/04/14/tox-tricks-and-patterns/#environment-reuse

**Comment by stsewd** [comment]
yeah, setting manually the env dir should do it :+1:

**Comment by stsewd** [comment]
> I don't have great feedback here, but I'm mostly worried about the team running tests locally. I almost always just run a subset of tests locally (tox -e py3 -- -s -k test_front_slash_url readthedocs/proxito/tests/test_middleware.py) -- so the defaults don't matter much to me.

I do run all tests before sending a PR, so I don't need to wait for the CI to complete to fix something, but yeah, mostly I do -k

**Comment by humitos** [comment]
>  I almost always just run a subset of tests locally (`tox -e py3 -- -s -k test_front_slash_url readthedocs/proxito/tests/test_middleware.py`) -- so the defaults don't matter much to me.

For the record: this runs just one test but it executes `pytest` twice; one for `-m "not proxito"` and another one for `-m proxito`, taking some extra time by collecting the tests twice from files that it's known they are not going to run.

> There must be a way to reuse envs. Does this work? [blog.ionelmc.ro/2015/04/14/tox-tricks-and-patterns/#environment-reuse](https://blog.ionelmc.ro/2015/04/14/tox-tricks-and-patterns/#environment-reuse)

This is probably a better approach that does not have this problem. I'd vote to go in this direction unless there are other concerns that weren't talked about yet.

File: tox.ini
Comment (humitos) [inline tox.ini:56]: Why do we need to make it explicit for all the testenv now? Isn't easier to just specify `proxito` and `py36` as equal instead?

File: .circleci/config.yml
Comment (humitos) [inline .circleci/config.yml:16]:
```suggestion
      - run: tox -e py36,proxito,search,codecov
```

This should call the tests for `search` as well.

File: docs/development/tests.rst
Comment (humitos) [inline docs/development/tests.rst:92]: Why are we removing the `search` mark? Didn't we agree that we were creating `proxito` and `search` tox testenv by sharing the environment itself to speed up?

I definitely don't want to run `search` tests locally when running my app tests. I don't spin search when doing docker up because it kills my computer; so all these tests fail when picked by pytest.

**Review by humitos — CHANGES_REQUESTED** [review]
The updates look good!

I'd like to see `-e search` as we talked about at https://github.com/readthedocs/readthedocs.org/pull/8116#issuecomment-856346799 before merging this PR. Is there anything blocking us from doing that?

There is no need to remove the `search` marks from the code. I want to have a way to skip them all because it's not trivial executing them: it requires the `search` container that consumes a lot of resources. I don't want to run search tests by default and I want to have a specific way to make them run isolated.

File: tox.ini
Comment (stsewd) [reply tox.ini:56]: the override comes from the main environment "py36", all other environments inherit from this one.

File: docs/development/tests.rst
Comment (stsewd) [reply docs/development/tests.rst:92]: I think we all agree to be able to run tests with `-k`, marks are used only for different configuration files. I'm -1 on having another environment, as it requires more typing to run the whole suite, and requires us tracking search tests, I think we don't even track those in .com.

**Comment by stsewd** [comment]
I'm still +1 on this refactor, looks like the only thing left is about search tests https://github.com/readthedocs/readthedocs.org/pull/8116#discussion_r650752218 (which are already being run by default on .com).

**Comment by ericholscher** [comment]
Seems like we've been talking about a few different test refactors. We should probably have a small proposal for moving forward on something that improves the situation. I generally agree with @humitos that our tests feel very complex, so I'd love to find a way to use markers more if possible, or some way to simplify how we run things.

**Comment by humitos** [comment]
It seems this is not gonna happen. Closing for now. We can recover this work once we decide to effectively do it

**Pull request closed without merging**
